c10/util/complex.h

```diff
@@ -0,0 +1,477 @@
+  constexpr complex<T> &operator =(const std::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  C10_HOST_DEVICE complex<T> &operator =(const thrust::complex<U> &rhs) {
+    storage[0] = rhs.real();
+    storage[1] = rhs.imag();
+    return static_cast<complex<T> &>(*this);
+  }
+#endif
+
+  template<typename U>
+  explicit constexpr operator std::complex<U>() const {
+    return std::complex<U>(std::complex<T>(real(), imag()));
+  }
+
+#if defined(__CUDACC__) || defined(__HIPCC__)
+  template<typename U>
+  explicit operator thrust::complex<U>() const {
+    return thrust::complex<U>(thrust::complex<T>(real(), imag()));
+  }
+#endif
+
+  constexpr T real() const {
+    return storage[0];
+  }
+  constexpr void real(T value) {
+    storage[0] = value;
+  }
+  constexpr T imag() const {
+    return storage[1];
+  }
+  constexpr void imag(T value) {
+    storage[1] = value;
+  }
+};
+
+
+template<>
+struct alignas(8) complex<float>: public complex_common<float> {
+  using complex_common<float>::complex_common;
+  constexpr complex(): complex_common() {}; // needed by CUDA 9.x
+  explicit constexpr complex(const complex<double> &other);
+  using complex_common<float>::operator=;
+};
+
+template<>
+struct alignas(16) complex<double>: public complex_common<double> {
+  using complex_common<double>::complex_common;
+  constexpr complex(): complex_common() {}; // needed by CUDA 9.x
+  constexpr complex(const complex<float> &other);
+  using complex_common<double>::operator=;
+};
+
+constexpr complex<float>::complex(const complex<double> &other): complex_common(other.real(), other.imag()) {}
+constexpr complex<double>::complex(const complex<float> &other): complex_common(other.real(), other.imag()) {}
+
+namespace complex_literals {
+
+constexpr complex<float> operator"" _if(long double imag) {
+  return complex<float>(float(), static_cast<float>(imag));
+}
+
+constexpr complex<double> operator"" _id(long double imag) {
+  return complex<double>(double(), static_cast<double>(imag));
+}
+
+} // namespace complex_literals
+
+} // namespace c10
+
+template<typename T>
+constexpr c10::complex<T> operator+(const c10::complex<T>& val) {
+  return val;
+}
+
+template<typename T>
+constexpr c10::complex<T> operator-(const c10::complex<T>& val) {
+  return c10::complex<T>(-val.real(), -val.imag());
+}
+
+template<typename T>
+constexpr c10::complex<T> operator+(const c10::complex<T>& lhs, const c10::complex<T>& rhs) {
+  c10::complex<T> result = lhs;
+  return result += rhs;
+}
+
+template<typename T>
+constexpr c10::complex<T> operator+(const c10::complex<T>& lhs, const T& rhs) {
+  c10::complex<T> result = lhs;
+  return result += rhs;
+}
+
+template<typename T>
+constexpr c10::complex<T> operator+(const T& lhs, const c10::complex<T>& rhs) {
+  c10::complex<T> result = rhs;
+  return result += lhs;
```

Comment:
Floating point is commutative so I guess this is OK, but it does make me a little nervous.

Reply:
I changed it to `return c10::complex<T>(lhs + rhs.real(), rhs.imag());`